src/update_agent/mod.rs

```diff
@@ -238,6 +248,53 @@ impl UpdateAgent {
     }
 }
 
+/// Attempt to broadcast msg to all sessions registered in systemd's login manager.
+/// Returns count of sessions broadcasted to.
+fn broadcast(msg: &str) -> Fallible<u8> {
+    // Get list of sessions with users logged in with `loginctl`.
+    let cmdrun = std::process::Command::new("loginctl")
+        .arg("list-sessions")
+        .arg("--output=json")
+        .output()
+        .with_context(|_| "failed to run 'loginctl' binary")?;
+
+    if !cmdrun.status.success() {
+        bail!(
+            "loginctl failed to list current sessions:\n{}",
+            String::from_utf8_lossy(&cmdrun.stderr)
+        );
+    }
+
+    let sessions: Vec<SessionsJSON> = serde_json::from_slice(&cmdrun.stdout)?;
```

Comment:
It would be good to split the logic up to here to its own function, and pass the sessions vector as input to the rest of the logic.